### PR TITLE
Download dependencies from json file

### PR DESCRIPTION
**Pull request by shayAlaluf**
added option to download dependencies from json file (package.json)

### CODE REVIEW

File: bin/npmDownload.js [hunk 1]
```diff
@@ -10,15 +11,16 @@ let program = require('commander'),
...
     .option('-e, --devDependencies', 'Download dev dependencies.')
     .option('-a, --allVersions', 'Download all versions of each package.')
     .option('-o, --output [directory]', 'The output [directory].')
     .option('-7, --zipIt', '7Zips the downloaded files.')
     .parse(process.argv);
 
-if (!program.packages) {
-    winston.error("No npmPackage were given!");
+if (!program.packages && !program.packagesFromJsonFile) {
+    winston.error("No npmPackage of path to jsonFile were given!");
```
Comment (gotenxds) [inline bin/npmDownload.js]: Typo ? should be "No npmPackage or path to jsonFile were given!"

File: bin/npmDownload.js [hunk 2]
```diff
@@ -10,15 +11,16 @@ let program = require('commander'),
...
     .option('-d, --dependencies', 'Download dependencies.')
     .option('-e, --devDependencies', 'Download dev dependencies.')
     .option('-a, --allVersions', 'Download all versions of each package.')
     .option('-o, --output [directory]', 'The output [directory].')
     .option('-7, --zipIt', '7Zips the downloaded files.')
     .parse(process.argv);
 
-if (!program.packages) {
-    winston.error("No npmPackage were given!");
+if (!program.packages && !program.packagesFromJsonFile) {
```
Comment (gotenxds) [inline bin/npmDownload.js:22]: Shouldn't this be an OR ?

File: bin/npmDownload.js [hunk 3]
```diff
@@ -27,6 +29,7 @@ if (!program.output) {
     process.exit();
 }
 
+let npmPackages = program.packagesFromJsonFile ? program.packagesFromJsonFile : program.packages;
```
Comment (gotenxds) [inline bin/npmDownload.js]: What if both arguments where given ? merge into one list.

File: coercions/packagesFromJsonFileCoercion.js [hunk 4]
```diff
@@ -0,0 +1,27 @@
+'use strict';
+let fs = require('fs'),
+    packages = require('./packagesCoercion');
+
+module.exports = function (jsonFilePath) {
+    const fileContent = JSON.parse(fs.readFileSync(jsonFilePath, "utf8"));
```
Comment (gotenxds) [inline coercions/packagesFromJsonFileCoercion.js]: I don't love the fact you are using sync here but I guess It's fine for now.

File: coercions/packagesFromJsonFileCoercion.js [hunk 5]
```diff
@@ -0,0 +1,27 @@
...
+    const dependencies = attachNameAndVersion(fileContent.dependencies);
+    const devDependencies = attachNameAndVersion(fileContent.devDependencies);
```
Comment (gotenxds) [inline coercions/packagesFromJsonFileCoercion.js]: You should check if the download dev-dependencies flag is true, its false by default.

File: coercions/packagesFromJsonFileCoercion.js [hunk 6]
```diff
@@ -0,0 +1,27 @@
...
+        let version = jsonObject[dependency];
+
+        if (version[0] !== '@') {
+            version = `@${version}`;
+        }
+
+        orderedDependencies += `${dependency}${version} `;
+    }
+
+    return orderedDependencies;
```
Comment (gotenxds) [inline coercions/packagesFromJsonFileCoercion.js]: `return orderedDependencies.trimRight()`
This should make no difference, better safe then sorry.

**Review by gotenxds — CHANGES_REQUESTED** [review]
Fix this please.

File: coercions/packagesFromJsonFileCoercion.js [hunk 7]
```diff
@@ -0,0 +1,27 @@
+'use strict';
+let fs = require('fs'),
+    packages = require('./packagesCoercion');
+
+module.exports = function (jsonFilePath) {
```
Comment (gotenxds) [inline coercions/packagesFromJsonFileCoercion.js]: arrow function.

File: coercions/packagesFromJsonFileCoercion.js (hunk 5 above)
Comment (shayAlaluf) [reply coercions/packagesFromJsonFileCoercion.js]: dev-dependencies flag will dowload all the dev dependencies of each dependency from the json file if set to true.
would u rather to set another flag to dowload dev dependencies from the json file?

File: bin/npmDownload.js (hunk 2 above)
Comment (shayAlaluf) [reply bin/npmDownload.js:22]: no. if nither of them appear in the command, then it's an error.
you can call: -j "[path-to-json-file]" or: -p ...

**Pull request merged**